src/librekey/key_store_g10.c

```diff
@@ -441,13 +527,305 @@ parse_seckey(pgp_keydata_key_t *keydata, s_exp_t *s_exp, pgp_pubkey_alg_t alg)
+
+    if (!parse_sexp(&s_exp, &bytes, &length)) {
+        pgp_forget(decrypted, key->key.seckey.encrypted_len);
+        free(decrypted);
+        return NULL;
+    }
+
+    // ignore padding data
+
+    pgp_forget(decrypted, key->key.seckey.encrypted_len);
+    free(decrypted);
+
+    if (s_exp.sub_s_expc == 0) {
+        destroy_s_exp(&s_exp);
+        (void) fprintf(stderr, "Hasn't got sub s-exp with key data.\n");
+        return NULL;
+    }
+
+    seckey = calloc(1, sizeof(*seckey));
+    if (seckey == NULL) {
+        destroy_s_exp(&s_exp);
+        (void) fprintf(stderr, "can't allocate memory\n");
+        return NULL;
+    }
+
+    seckey->pubkey = key->key.seckey.pubkey;
+
+    if (!parse_seckey(seckey, &s_exp.sub_s_exps[0], seckey->pubkey.alg, true)) {
+        destroy_s_exp(&s_exp);
+        return NULL;
+    }
+
+    destroy_s_exp(&s_exp);
+
+    return seckey;
+}
+
+static bool
+parse_protected_seckey(pgp_seckey_t *seckey, s_exp_t *s_exp)
+{
+    const format_info *format;
+
+    s_exp_t *var = lookup_variable(s_exp, "protected");
+    if (var == NULL) {
+        return NULL;
+    }
+
+    format = parse_format((const char *) var->blocks[1].bytes, var->blocks[1].len);
+    if (format == NULL) {
+        fprintf(stderr,
+                "Unsupported protected mode: '%.*s'\n",
+                (int) var->blocks[1].len,
+                var->blocks[1].bytes);
+        return false;
+    }
+
+    seckey->alg = format->cipher;
+    seckey->cipher_mode = format->cipher_mode;
+    seckey->hash_alg = format->hash_alg;
+
+    if (var->sub_s_expc != 1) {
+        fprintf(stderr, "Wrong count of sub-level s-exp: %d, should be 1\n", var->sub_s_expc);
+        return false;
+    }
+
+    s_exp_t *encrypted = &var->sub_s_exps[0];
+
+    if (encrypted->sub_s_expc != 1 && encrypted->blockc != 2) {
+        fprintf(stderr,
+                "Wrong count of sub-level s-exp: %d or blocks: %d, should be 1 and 2\n",
+                encrypted->sub_s_expc,
+                encrypted->blockc);
+        return false;
+    }
+
+    s_exp_t *alg = &encrypted->sub_s_exps[0];
+
+    if (strncmp("sha1", (const char *) alg->blocks[0].bytes, alg->blocks[0].len) != 0) {
+        fprintf(stderr,
+                "Wrong hashing algorithm, should be sha1 but %.*s\n",
+                (int) alg->blocks[0].len,
+                alg->blocks[0].bytes);
+        return false;
+    }
+
+    seckey->hash_alg = PGP_HASH_SHA1;
+    seckey->s2k_usage = PGP_S2KU_ENCRYPTED;
+    seckey->s2k_specifier = PGP_S2KS_ITERATED_AND_SALTED;
+
+    if (alg->blocks[1].len != PGP_SALT_SIZE) {
+        fprintf(stderr,
+                "Wrong salt size, should be %d but %d\n",
+                PGP_SALT_SIZE,
+                (int) alg->blocks[1].len);
+        return false;
+    }
+
+    memcpy(seckey->salt, alg->blocks[1].bytes, alg->blocks[1].len);
+    seckey->s2k_iterations = block_to_unsigned(&alg->blocks[2]);
+    if (seckey->s2k_iterations == UINT8_MAX) {
```

Comment:
Hmm... I think it should it be seckey->s2k_iterations == UINT_MAX instead of UINT8_MAX (see line 294)